### ISSUE TITLE
Reproducibility of the paper results with pretrained model

### ISSUE THREAD

**Issue by annwhoorma** [jianlin-cheng/CryoSegNet#1]
Hi there!

I've read your [paper](https://www.biorxiv.org/content/10.1101/2023.10.02.560572v1.full.pdf) and was amazed by your outcomes and your idea for the model architecture. I tried running the pretrained model on several EMPIAR classes:

- EMPIAR-10671 (from the train data you provided)
- EMPIAR-10345 (from the test data you provided)
- EMPIAR-10146 (from [EMPIAR website](https://www.ebi.ac.uk/empiar/EMPIAR-10146/), this class has 20 micrographs in _.mrc_ format)

However, visually, the results' precision didn't correspond to the paper's results. I double-checked that I used pretrained weights, but maybe I missed something else. I ran the prediction on Colab environment, and you can see/copy&run it by following [this link](https://colab.research.google.com/drive/1vyde4v3I9Jo0ft0KHWSwZ4rEHJcORgUm?usp=sharing).

Below, I will attach input, code and output for some cases.

**EMPIAR-10671**

This one belong to **train** data, so it should be downloaded, as well as model checkpoints for U-Net and SAM.

```python
!python CryoSegNet/predict_new_data_jpg.py \
                                --my_dataset_path '/content/train_dataset/10671/train/images' \
                                --cryosegnet_checkpoint '/content/pretrained_models/cryosegnet.pth' \
                                --sam_checkpoint '/content/pretrained_models/sam_vit_h_4b8939.pth' \
                                --output_path '/content/CryoSegNet/output'
```
A couple of the outputs that indicate a lot of overlapping between the circles and false positives:

| PS0114 RAMP1-CGRP_110-5_0001_Jan16_12 35 00 | PS0114 RAMP1-CGRP_110-5_0003_Jan16_12 35 15 |
| ------------  | ------------ |
| ![predicted_PS0114 RAMP1-CGRP_110-5_0001_Jan16_12 35 00_result](https://github.com/jianlin-cheng/CryoSegNet/assets/34862867/b8d15338-e5ec-4cb0-9dd2-ca8bd4588a62)| ![predicted_PS0114 RAMP1-CGRP_110-5_0003_Jan16_12 35 15_result](https://github.com/jianlin-cheng/CryoSegNet/assets/34862867/02213507-e6f2-4668-a11d-a0fd284c7a54)|

**EMPIAR-10345**

This one belong to **test** data, so it should be downloaded.
I was especially interested in this one, as the paper claims to achieve 1.58 A for resolution. However, the results still contains a lot of non-particles picked up, I am attaching two examples below:

```python
!python CryoSegNet/predict_new_data_jpg.py \
                                --my_dataset_path '/content/test_dataset/10345/images' \
                                --output_path '/content/CryoSegNet/output' \
                                --cryosegnet_checkpoint '/content/pretrained_models/cryosegnet.pth' \
                                --sam_checkpoint '/content/pretrained_models/sam_vit_h_4b8939.pth' \
```

| 18jam15a_0155_ali_DW | 18jam15a_0236_ali_DW |
| ------------  | ------------ |
| ![predicted_18jam15a_0155_ali_DW_result](https://github.com/jianlin-cheng/CryoSegNet/assets/34862867/8e6fad19-82d0-4c69-ae28-056e0e742e83) | ![predicted_18jam15a_0236_ali_DW_result](https://github.com/jianlin-cheng/CryoSegNet/assets/34862867/a16f829b-8968-45a4-a6d9-de68ed0440b8) |

**EMPIAR-10146**

I have a few motion-corrected apoferritin images in _.mrc_ format, and one of them can be downloaded via this [link](https://drive.google.com/file/d/1Xrkie2NyxE-6GM25E27_f2jTHW_Dn2L2/view?usp=sharing) from Google Drive. You can see it in the first column.

I tried to use pretrained CryoSegNet to detect the proteins, but didn't succeed, even though the picture contains very little noise.

```python
!python CryoSegNet/predict_new_data_mrc.py \
                    --my_dataset_path '/content/mydata/' \
                    --output_path '/content/CryoSegNet/output' \
                    --cryosegnet_checkpoint '/content/pretrained_models/cryosegnet.pth' \
                    --sam_checkpoint '/content/pretrained_models/sam_vit_h_4b8939.pth' \
```

The picture seemed to undergo some sort of denoising, almost blurring, and was flipped. 
**Question**: Could it be that the mask was predicted for the non-flipped image? I tried to put the final mask on top of the original image, and it still isn't precise.
| original | predicted |
| ------------  | ------------ |
| ![image](https://github.com/jianlin-cheng/CryoSegNet/assets/34862867/e7683612-4908-4d72-8d8c-1b3449abc227) | <img src="https://github.com/jianlin-cheng/CryoSegNet/assets/34862867/947f6acb-6ea1-4576-a0de-c3fafbb77221" alt="drawing" width="380"/> |

**Final part**

1. With these screenshots and my Colab notebook, could you please let me know if you noticed a mistake in my steps that led to the incorrect results? Otherwise, what else could be a possible reason for such imprecise results that I obtained?
2. Why is the resulting image with the predicted mask is flipped?

I appreciate that you took time to read it and hoping for a reply/chat!

**Comment by rajangyawali** [comment]
Hi,

1. Regarding 10671, yes we can see some overlapping between false positives and the predicted circles. We can eliminate these using 2D classification job and Select 2D job in the CryoSPARC. The star file predicted by CryoSegNet can be imported in CryoSPARC for further post processing.
2.  Regarding 10345, we may feel like some of the non-particles being picked up because we have used denoised version of the micrographs  which may have enhanced the contrast of some background noise. The deEven if the non-particles are picked up, we can use the CryoSPARC to eliminate such particles as mentioned in 1. 
3.  Regarding 10146, I will try with the images you have provided and let you know once it is fixed.

Thank you so much for your interest in this model. We are working to address all the issues that have been raised to make this tool more robust. Researchers and users like you will always help to drive the research in positive direction, We appreciate a lot for your time in trying our model.

**Comment by annwhoorma** [comment]
Thanks for a quick reply!

As for the 1st point, the Table 2 from the paper provides the same resolution of 3D density maps for EMPIAR-10028 protein with and without Select2D. It led me to an assumption that the model would perform very well on this protein even without Select2D due to low false positives count, so I ran the model on some images from EMPIAR-10028, but the resulting images still contain a lot of false positives and very little true positives. I am attaching a few below.

![predicted_076_result](https://github.com/jianlin-cheng/CryoSegNet/assets/34862867/cbf848e4-aa7d-45e6-90a2-114b11cb01f0)
![predicted_091_result](https://github.com/jianlin-cheng/CryoSegNet/assets/34862867/5a94034e-529f-4f28-8668-7d84b65a1fd9)
![predicted_343_result](https://github.com/jianlin-cheng/CryoSegNet/assets/34862867/cecd6cf8-c2df-46fe-8479-ddebf54da0a0)

What could be the reason for such poor results in my case?

**Comment by rajangyawali** [comment]
Hi,

Thank you for helping us to make our code more robust.

These results  were because we applied denoising on images which were already denoised. Could you please try again with just python predict.py (make sure to pull the recent code).

For Example: python predict.py --empiar_id 10028

The results should look like this:
![predicted_061_result](https://github.com/jianlin-cheng/CryoSegNet/assets/15334991/52e453de-faec-4699-8427-e80dfc4c985b)
![predicted_062_result](https://github.com/jianlin-cheng/CryoSegNet/assets/15334991/c506ead8-77ef-483b-8cf3-8848679d74f9)
![predicted_063_result](https://github.com/jianlin-cheng/CryoSegNet/assets/15334991/dfc1b5fe-ff94-46c4-a52c-13f7c00dfc47)

**Comment by annwhoorma** [comment]
Thanks! Indeed, that's right :) I also ran it for EMPIAR-10345 the same way and got much better results than before.

As you might remember, I opened one more issue about lack of denoising in _predict.py_ script, and that issue is closed by now. The resolution of that issue led to adding denoising step in _predict.py_, and then this issue led to removing this step again. Also, it remains confusing that _predict.py_ and scripts for predicting for new data have different pipelines. I'd suggest to maybe add it in the README, so that no one gets confused :)
There are also other inconsistencies between the scripts, e.g. _predict.py_ has `np.rot90` call commented, while _generate_starfile_new_data_mrc.py_ has it in the script. Could it also be the cause for wrong predictions?

---
I've updated this comment several times because all these times I found answers to my questions myself eventually, by looking through the repository
